prow/cmd/deck/main.go

```diff
@@ -142,6 +144,52 @@ func staticHandlerFromDir(dir string) http.Handler {
 	return gziphandler.GzipHandler(handleCached(http.FileServer(http.Dir(dir))))
 }
 
+var (
+	deckMetrics = struct {
+		httpRequestDuration *prometheus.HistogramVec
+	}{
+		httpRequestDuration: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name:    "deck_http_request_duration_seconds",
+				Help:    "http request duration seconds for deck server",
+				Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10},
+			},
+			[]string{"path", "method", "status"},
+		),
+	}
+)
+
+type traceResponseWriter struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (trw *traceResponseWriter) WriteHeader(code int) {
+	trw.statusCode = code
+	trw.ResponseWriter.WriteHeader(code)
+}
+
+func NewTraceResponseWriter(w http.ResponseWriter) *traceResponseWriter {
+	// Initialize the status to 200 in case WriteHeader is not called
+	return &traceResponseWriter{w, http.StatusOK}
+}
+
+func init() {
+	prometheus.MustRegister(deckMetrics.httpRequestDuration)
+}
+
+func traceHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		t := time.Now()
+		trw := NewTraceResponseWriter(w)
+		h.ServeHTTP(trw, r)
+		latency := time.Since(t)
+		deckMetrics.httpRequestDuration.With(
+			prometheus.Labels{"path": r.URL.Path, "method": r.Method, "status": strconv.Itoa(trw.statusCode)}).
```

Comment:
I think we will need to do some prefixing on this path -- for instance, all of the Spyglass pages will have lots of specifics to the job/pr/etc that is being requested, but we want to distinguish between handlers in this file, really

Reply:
So maybe we want to record:

```
/tide.js
/tide-history.js
/plugin-help.js
/data.js
/prowjobs.js
/pr-data.js
/log
/rerun
/spyglass/static/*
/spyglass/lens/*
/view/*
/job-history/*
/pr-history/*
```

@Katharine might have thoughts as well